x/bank/keeper/keeper.go

```diff
@@ -419,6 +419,33 @@ func (k BaseKeeper) BurnCoins(ctx context.Context, moduleName string, amounts sd
 	return nil
 }
 
+// AddBalance is low level api to update balance directly, mainly used by evm integration,
+// caller should make sure the total supply of the denom not changed and the account exists.
+// it's for evm integration, call at your own risk.
+// it emits mint event.
+func (k BaseKeeper) AddBalance(ctx context.Context, addr sdk.AccAddress, coin sdk.Coin) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	// emit mint event
+	sdkCtx.EventManager().EmitEvent(
+		types.NewCoinMintEvent(addr, sdk.NewCoins(coin)),
+	)
+	return k.addCoin(ctx, addr, coin)
```

Comment:
might handle add sub coin first, we don't need emit event if error right?

Reply:
it's always executed in a cache context, it's atomic anyway.

Reply:
reordered, it make sense in the sense of efficiency